### DISCUSSION TITLE
Fastify methods conflicting with `@typescript-eslint/no-floating-promises` rule

### DISCUSSION THREAD

**Discussion by leonsilicon** [fastify/fastify#3849] [Q&A]
Hi all! I was wondering if there's a more ergonomic way to use fastify methods with `@typescript-eslint/no-floating-promises` rule. A lot of fastify's reply methods return the original reply object, and the original object has a then property on it which triggers this eslint warning. The problem is, awaiting `reply` causes the server to hang because it will only resolve once the reply has been sent. Right now, I have to add the void operator (`void context.reply.setCookie(sessionTokenCookie, sessionToken);`) every time I use one of `reply`'s methods, which is a bit tedious (especially when I forget that I have to use `void` and use `await` instead and then have to debug why my server is hanging)

![Screen_Shot_2022-04-20_at_3 36 36_PM](https://user-images.githubusercontent.com/36966635/164915293-836dac76-b8c8-412e-aabb-4fa15ea73280.png)

Optimally, the `reply` object shouldn't have a `.then()` and instead should be a property on the object like `reply.sentPromise`, but I imagine this would be a pretty big breaking change.

**Comment by mcollina** [comment]
I think we should document to disable that rule.

I'm also happy to _add_ a `sentPromise` and remove `.then()`  from the type definitions. Something event better would be to tell tslint that `Reply` is not a promise and it's ok to not await it.

cc @fastify/typescript

**Reply by leonsilicon** [reply]
I think removing `.then()` from the type definitions and adding a property like `sentPromise` would be a good way to solve this 👍

As far as I know, I don't think there's currently a way to tell typescript-eslint that `Reply` isn't a promise, and it looks like the typescript-eslint team prefers that `void` is explicitly used in these scenarios (https://github.com/typescript-eslint/typescript-eslint/issues/4791)

**Reply by fox1t** [reply]
I like that `Reply` is a thenable, and I would not remove the `then` method. 
On the other hand, unfortunately, if you use `no-floating-point` rule, you are supposed to handle any thenable (and not only "real" Promises). 
Using the `void` is the correct thing when `o-floating-promises` is turned on.  :/

**Reply by leonsilicon** [reply]
I think adding `void` for `Reply` might make the code a bit confusing in some scenarios, as it makes methods like `reply.setCookie('foo', 'bar')` which return the original reply look like asynchronous functions. Not sure if it's possible, but maybe there's something we can do with the types using interface augmentation or something similar to have an optional importable type definition that gets rid of the `.then()` type and adds the `sentPromise` property?

**Comment by fox1t** [comment]
I think that the issue here is on the eslint side. They use this function to detect the "floating promises" https://github.com/typescript-eslint/typescript-eslint/blob/6c3816b3831e6e683c1a7842196b34248803d69b/packages/eslint-plugin/src/rules/no-floating-promises.ts#L180

In reality, they are searching for "thenables". So I would suggest opening an issue on their side so that they can add a new flag, "ignoreThenable" (or something like that), to control the detection and report only the Promises. 

Note: Understanding if something is a promise or a thenable is an entirely different challange XD

**Reply by leonsilicon** [reply]
Just found this issue https://github.com/typescript-eslint/typescript-eslint/issues/2640 which mentions the problem with Fastify's `Reply`, and it unfortunately looks like they couldn't determine an effective solution to the problem.
